sorted_list_dict.py

```diff
@@ -12,20 +12,36 @@ def __eq__(self, other):
     def __str__(self) -> str:
         return self.__list.__str__()
 
+    def __len__(self) -> int:
+        return self.__list.length()
+
     def __find_gte_or_length(self, key: str) -> int:
         '''
         gte means greater or equal
         '''
-        # TODO: binary search
+        # binary search
+
+        left = 0  # left border
+        right = len(self.__list) - 1  # right border
+        middle = 0
 
-        index = 0
+        while left <= right:
+            middle = (left + right)//2
 
-        while index < self.__list.length() and self.__list[index][0] < key:
-            index += 1
+            if self.__list[middle][0] == key:
+                return middle
+            elif key > self.__list[middle][0]:
+                left += 1
+            else:
+                right -= 1
 
-        return index
+        # check if List is empty to avoid exception
+        return middle if len(self.__list) == 0 or key < self.__list[middle][0] else middle + 1
```

Comment:
Current implementation:
While the `left` one is less than or equal to the `right`, we are trying to find matches by `key`.
If the loop has ended, then we have an empty list or there is no desired element in the list.
In the second case, we return `middle`, because the `key` standing at this position in the list is greater than the `key`. which we want to insert.
Otherwise, we insert into the `middle + 1` position, because the `key` is greater than the `key` that stands on the `middle` index.